Agents/HistoricalPumpDataInstantiation/HistoricalPumpDataInstantiationAgent/src/main/java/uk/ac/cam/cares/jps/agent/TimeSeriesClientDecorator.java

```diff
@@ -0,0 +1,173 @@
+ *
+ * @author qhouyee
+ */
+class TimeSeriesClientDecorator {
+    // Logger for reporting info/errors
+    private static final Logger LOGGER = LogManager.getLogger(HistoricalPumpDataInstantiationAgent.class);
+    private TimeSeriesClient<Instant> tsClient;
+    private RemoteRDBStoreClient rdbClient;
+    public final String timeUnit = Instant.class.getSimpleName(); // The time unit used by the Excel Workbook
+    private final String timeKey;
+
+    /**
+     * Standard constructor
+     */
+    protected TimeSeriesClientDecorator(String timeHeader) {
+        this.timeKey = timeHeader;
+    }
+
+    /**
+     * Setter for the time series client.
+     *
+     * @param tsClient The time series client to use.
+     */
+    protected void setTsClient(TimeSeriesClient<Instant> tsClient) {
+        this.tsClient = tsClient;
+    }
+
+    /**
+     * Get the RDB client.
+     */
+    protected RemoteRDBStoreClient getRDBClient() {
+        return this.rdbClient;
+    }
+
+    /**
+     * Set a new RDB client.
+     *
+     * @param url      The RDB JDBC URL.
+     * @param username The RDB username.
+     * @param password The RDB password.
+     */
+    protected void setRDBClient(String url, String username, String password) {
+        this.rdbClient = new RemoteRDBStoreClient(url, username, password);
+    }
+
+    /**
+     * Initializes all time series if they do not exist using the time series client.
+     *
+     * @param excelReadings Excel readings parsed from the Excel Workbook
+     * @param iriMappings   Mappings between measures' names and their corresponding data IRI.
+     */
+    protected void initializeTimeSeries(Map<String, List<?>> excelReadings, Map<String, String> iriMappings) throws SQLException {
+        List<String> excelHeaders = new ArrayList<>(iriMappings.keySet());
+        List<String> iris = new ArrayList<>(iriMappings.values());
+        // If IRIs do not have a time series linked, initialize the corresponding time series
+        if (timeSeriesDoesNotExist(iris)) {
+            LOGGER.debug("Time series does not exist on RDB. Generating time series templates...");
+            // Get the classes (datatype) corresponding to each measure's name needed for initialization
+            // excelReadings.get(header) returns array list, so add get method to get the value
+            List<Class<?>> classes = excelHeaders.stream()
+                    .map(header -> excelReadings.get(header).get(0).getClass())
+                    .collect(Collectors.toList());
+            try (Connection conn = rdbClient.getConnection()) {
+                tsClient.initTimeSeries(iris, classes, this.timeUnit, conn, TimeSeriesClient.Type.STEPWISECUMULATIVE, null, null);
+                LOGGER.debug(String.format("Initialized time series with the following IRIs: %s", String.join(", ", iris)));
+            } finally {
+                // Always close connection
+                rdbClient.getConnection().close();
```

Comment:
Since getConnection will open a new connection, this line is totally pointless. The try-with-resources block above should already close the connection automatically.